### ISSUE TITLE
Overriding interface method's return type via @method in implementation does not work

### ISSUE THREAD

**Issue by Wirone** [vimeo/psalm#999]
In one of our projects (Yii2) we have models based on `ActiveRecord`. Psalm is complaining:

```
ERROR: UndefinedMethod - 57:9 - Method ActiveQueryInterface::viatable does not exist
```
even if `ActiveRecord` has `@method` override in phpDoc and `ActiveQuery` should be considered as a result of `hasMany()`, so `viaTable()` is available. But Psalm considers `ActiveQueryInterface` as a result of `hasMany()`, as `BaseActiveRecord::hasMany()` defines.

I have `usePhpDocMethodsWithoutMagicCall` set to `true`, still there is an error.

Live example: https://getpsalm.org/r/7b12596d4c

**Comment by muglug** [comment]
the problem here is that `hasMany` is defined on `BaseActiveRecord`. That stops Psalm from triggering any of the magic method logic.

You could define the method on `ActiveRecord` like so: https://getpsalm.org/r/c145b479da

**Comment by Wirone** [comment]
But `ActiveRecord` is a part of Yii's core so I can't define `hasMany()` there, I could only make custom AR-base class just for that, which does not makes sense IMO.

**Comment by bramcordie** [comment]
I ran into a similar problems when using an `@method` annotation on a child class where the actual implementation is in the parent class.

My use-case is an abstract `Collection` class that has a `mergedWith(Collection $other): Collection` method. The classes extending `Collection` specify that this method returns a new instance of the child class. Psalm still thinks that an instance of both the parent or child class can be returned. PHPStorm does pick up the specified return type as expected.

**Comment by muglug** [comment]
ok, I think I might add some custom behaviour for when the code is not user-editable. You can always stub these methods, but I understand how that might be annoying.

**Comment by muglug** [comment]
Ok, given
```php
class A {
  public function foo(string $s) : string {
    return $s;
  }
}
```

Psalm will treat

```php
/** @method stdClass foo(stdClass $s) */
class B extends A {}
```

identically to

```php
class B extends A {
  /**
   * @param stdClass $s
   * @return stdClass
   */
  public function foo(string $s) : string {
    return parent::foo($s);
  }
}
```

**Comment by bramcordie** [comment]
I tried to apply your stubbing example but that does not work with how things are setup for my use-case. Here is what I'm trying to accomplish: https://getpsalm.org/r/98e1842a43.

**Comment by Rarst** [comment]
Also having problem with extending. I extend Slim's `App` class which has `getContainer()` doc blocked to return `ContainerInterface` but my subclass needs to hint concrete implementation `@method Container getContainer()` because it has methods I need that are not in the interface.

**Comment by weirdan** [comment]
>  I extend Slim's App class which has getContainer() doc blocked to return ContainerInterface but my subclass needs to hint concrete implementation @method Container getContainer()

That doesn't work with PHP native type system BTW: https://3v4l.org/LPaF6

**Comment by Rarst** [comment]
I know, but that captures de–facto state of application in my case. Upstream Slim does in fact return `Container`, I am not _changing_ anything about that behavior. Slim documents interface because container class _could_ be replaced with a different one, but this is not something I support in _my_ app for historical reasons.

So my choices are either `@var Container` every single `getContainer()` call across the whole app or `@method Container getContainer()` it centrally (which PhpStorm and phan support for me already).

**Comment by muglug** [comment]
@bramcordie you can accomplish that with https://getpsalm.org/r/b72dbbaada (adding `@return static`)

**Comment by muglug** [comment]
Thanks everyone for input - I think the solution here respects the majority of use cases, but LMK if there's more that you need.

**Comment by bramcordie** [comment]
@muglug returning static works as expected, thanks for following up! Cleaned up example: 
https://getpsalm.org/r/b7229f7302
